File: CS/GrowthWare.Framework/Model/Profiles/Base/MGroupRolePermissionSecurity.cs
using GrowthWare.Framework.Model.Enumerations;
using GrowthWare.Framework.Model.Profiles.Interfaces;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Collections.ObjectModel;

namespace GrowthWare.Framework.Model.Profiles.Base
{
    /// <summary>
    /// The MBaseSecurity is a abstract class that when inherited will add 4 types of roles
    /// to your class/object.  After you have inherited the class pass a data row to the SecurityInit sub 
    /// to populate the roles.
    /// </summary>
    /// <remarks>
    /// Currently there are 4 permission roles and they are Add, Edit, View, Delete.  
    /// If you would like to extend this class do so by inheriting this class and adding your
    /// own types of roles say Moderate if your writing some sort of formum type of class.
    /// Any of the objects you create should now inherit your class and will now have all of the 
    /// roles from this class as well as the ones for yours.
    ///</remarks>
    [Serializable(), CLSCompliant(true)]
    public abstract class MGroupRolePermissionSecurity : MProfile, IMGroupRolePermissionSecurity
    {
        private Collection<string> m_AssignedAddRoles = new Collection<string>();
        private Collection<string> m_AssignedDeleteRoles = new Collection<string>();
        private Collection<string> m_AssignedEditRoles = new Collection<string>();
        private Collection<string> m_AssignedViewRoles = new Collection<string>();

        private Collection<string> m_DerivedAddRoles = new Collection<string>();
        private Collection<string> m_DerivedDeleteRoles = new Collection<string>();
        private Collection<string> m_DerivedEditRoles = new Collection<string>();
        private Collection<string> m_DerivedViewRoles = new Collection<string>();

        private Collection<string> m_AddGroups = new Collection<string>();
        private Collection<string> m_DeleteGroups = new Collection<string>();
        private Collection<string> m_EditGroups = new Collection<string>();
        private Collection<string> m_ViewGroups = new Collection<string>();
        private string m_PermissionColumn = "PERMISSIONS_SEQ_ID";

        private string m_RoleColumn = "ROLE";
        private string m_GroupColumn = "Group";

        /// <summary>
        /// Retruns a comma seporated string given a Collection strings
        /// </summary>
        /// <param name="collectionOfStrings">Collecion</param>
        /// <returns>comma seportated string.</returns>
        private string getCommaSeportatedString(Collection<string> collectionOfStrings)
        {
            String mRetValue = String.Empty;
            if (collectionOfStrings != null)
            {
                if (collectionOfStrings.Count > 0)
                {
                    foreach (var item in collectionOfStrings)
                    {
                        mRetValue += item.ToString() + ",";
                    }
                }
            }
            if (mRetValue.Length > 0)
            {
                mRetValue = mRetValue.Substring(0, mRetValue.Length - 1);
            }
            return mRetValue;
        }

        /// <summary>
        /// Initialize overloads and calls base.Initialize to will populate the Add, Delete, Edit, and View role properties.
        /// </summary>
        /// <param name="profileDataRow">The detail row.</param>
        /// <param name="derivedRoles">The derived roles.</param>
        /// <param name="assignedRoles">The assigned roles.</param>
        /// <param name="groups">The groups.</param>
        virtual protected void Initialize(DataRow profileDataRow, DataRow[] derivedRoles, DataRow[] assignedRoles, DataRow[] groups)
        {
            base.Initialize(profileDataRow);
            this.setRolesOrGroups(m_DerivedAddRoles, derivedRoles, PermissionType.Add, m_RoleColumn);
            this.setRolesOrGroups(m_DerivedDeleteRoles, derivedRoles, PermissionType.Delete, m_RoleColumn);
            this.setRolesOrGroups(m_DerivedEditRoles, derivedRoles, PermissionType.Edit, m_RoleColumn);
            this.setRolesOrGroups(m_DerivedViewRoles, derivedRoles, PermissionType.View, m_RoleColumn);
            if (assignedRoles != null)
            {
                this.setRolesOrGroups(m_AssignedAddRoles, assignedRoles, PermissionType.Add, m_RoleColumn);
                this.setRolesOrGroups(m_AssignedDeleteRoles, assignedRoles, PermissionType.Delete, m_RoleColumn);
                this.setRolesOrGroups(m_AssignedEditRoles, assignedRoles, PermissionType.Edit, m_RoleColumn);
                this.setRolesOrGroups(m_AssignedViewRoles, assignedRoles, PermissionType.View, m_RoleColumn);
            }
            if (groups != null)
            {
                this.setRolesOrGroups(m_AddGroups, groups, PermissionType.Add, m_GroupColumn);
                this.setRolesOrGroups(m_DeleteGroups, groups, PermissionType.Delete, m_GroupColumn);
                this.setRolesOrGroups(m_EditGroups, groups, PermissionType.Edit, m_GroupColumn);
                this.setRolesOrGroups(m_ViewGroups, groups, PermissionType.View, m_GroupColumn);
            }

        }

        /// <summary>
        /// Return assigned roles associated with the "Add" permission.
        /// </summary>
        Collection<string> IMGroupRolePermissionSecurity.AssignedAddRoles
        {
            get { return m_AssignedAddRoles; }
        }

        /// <summary>
        /// Return roles associated with the "Add" permission.
        /// </summary>
        Collection<string> IMGroupRolePermissionSecurity.DerivedAddRoles
        {
            get { return m_DerivedAddRoles; }
        }

        /// <summary>
        /// Return assigned roles associated with the "Delete" permission.
        /// </summary>
        Collection<string> IMGroupRolePermissionSecurity.AssignedDeleteRoles
        {
            get { return m_AssignedDeleteRoles; }
        }

        /// <summary>
        /// Return roles associated with the "Delete" permission.
        /// </summary>
        Collection<string> IMGroupRolePermissionSecurity.DerivedDeleteRoles
        {
            get { return m_DerivedDeleteRoles; }
        }

        /// <summary>
        /// Return roles associated with the "Edit" permission.
        /// </summary>
        Collection<string> IMGroupRolePermissionSecurity.AssignedEditRoles
        {
            get { return m_AssignedEditRoles; }
        }

        /// <summary>
        /// Return roles associated with the "Edit" permission.
        /// </summary>
        Collection<string> IMGroupRolePermissionSecurity.DerivedEditRoles
        {
            get { return m_DerivedEditRoles; }
        }

        /// <summary>
        /// Return assigned roles associated with the "View" permission.
        /// </summary>
        Collection<string> IMGroupRolePermissionSecurity.AssignedViewRoles
        {
            get { return m_AssignedViewRoles; }
        }

        /// <summary>
        /// Return roles associated with the "View" permission.
        /// </summary>
        Collection<string> IMGroupRolePermissionSecurity.DerivedViewRoles
        {
            get { return m_DerivedViewRoles; }
        }

        /// <summary>
        /// Return groups associated with the "Add" permission.
        /// </summary>
        Collection<string> IMGroupRolePermissionSecurity.AddGroups
        {
            get { return m_AddGroups; }
        }

        /// <summary>
        /// Return groups associated with the "Delete" permission.
        /// </summary>
        Collection<string> IMGroupRolePermissionSecurity.DeleteGroups
        {
            get { return m_DeleteGroups; }
        }

        /// <summary>
        /// Return groups associated with the "Edit" permission.
        /// </summary>
        Collection<string> IMGroupRolePermissionSecurity.EditGroups
        {
            get { return m_EditGroups; }
        }

        /// <summary>
        /// Return groups associated with the "View" permission.
        /// </summary>
        Collection<string> IMGroupRolePermissionSecurity.ViewGroups
        {
            get { return m_ViewGroups; }
        }

        /// <summary>
        /// Converts the collection of AssignedGroups to a comma seporated string.
        /// </summary>
        /// <returns>String</returns>
        public string GetCommaSeporatedGroups(PermissionType permission)
        {
            switch (permission)
            {
                case PermissionType.Add:
                    return this.getCommaSeportatedString(m_AddGroups);
                case PermissionType.Delete:
                    return this.getCommaSeportatedString(m_DeleteGroups);
                case PermissionType.Edit:
                    return this.getCommaSeportatedString(m_EditGroups);
                case PermissionType.View:
                    return this.getCommaSeportatedString(m_ViewGroups);
                default:
                    return null;
            }
        }

        /// <summary>
        /// Converts the collection of AssignedGroups to a comma seporated string.
        /// </summary>
        /// <returns>String</returns>
        public string GetCommaSeporatedAssingedRoles(PermissionType permission)
        {
            switch (permission)
            {
                case PermissionType.Add:
                    return this.getCommaSeportatedString(m_AssignedAddRoles);
                case PermissionType.Delete:
                    return this.getCommaSeportatedString(m_AssignedDeleteRoles);
                case PermissionType.Edit:
                    return this.getCommaSeportatedString(m_AssignedEditRoles);
                case PermissionType.View:
                    return this.getCommaSeportatedString(m_AssignedViewRoles);
                default:
                    return null;
            }
        }

        /// <summary>
        /// Represents the permission column name.
        /// </summary>
        public string PermissionColumn
        {
            get { return m_PermissionColumn; }
            set { m_PermissionColumn = value.Trim(); }
        }

        /// <summary>
        /// Represents the role column name.
        /// </summary>
        public string RoleColumn
        {
            get { return m_RoleColumn; }
            set { m_RoleColumn = value.Trim(); }
        }

        /// <summary>
        /// Sets the groups.
        /// </summary>
        /// <param name="commaSeporatedGroups">The comma seporated groups.</param>
        /// <param name="permission">The permission.</param>
        public void SetGroups(string commaSeporatedGroups, PermissionType permission)
        {
            switch (permission)
            {
                case PermissionType.Add:
                    this.setRolesOrGroups(m_AddGroups, commaSeporatedGroups);
                    break;
                case PermissionType.Delete:
                    this.setRolesOrGroups(m_DeleteGroups, commaSeporatedGroups);
                    break;
                case PermissionType.Edit:
                    this.setRolesOrGroups(m_EditGroups, commaSeporatedGroups);
                    break;
                case PermissionType.View:
                    this.setRolesOrGroups(m_ViewGroups, commaSeporatedGroups);
                    break;
                default:
                    break;
            }
        }

        /// <summary>
        /// Sets the assigned roles.
        /// </summary>
        /// <param name="commaSeporatedRoles">The comma seporated roles.</param>
        /// <param name="permission">The permission.</param>
        public void SetAssignedRoles(string commaSeporatedRoles, PermissionType permission)
        {
            switch (permission)
            {
                case PermissionType.Add:
                    this.setRolesOrGroups(m_AssignedAddRoles, commaSeporatedRoles);
                    break;
                case PermissionType.Delete:
                    this.setRolesOrGroups(m_AssignedDeleteRoles, commaSeporatedRoles);
                    break;
                case PermissionType.Edit:
                    this.setRolesOrGroups(m_AssignedEditRoles, commaSeporatedRoles);
                    break;
                case PermissionType.View:
                    this.setRolesOrGroups(m_AssignedViewRoles, commaSeporatedRoles);
                    break;
                default:
                    break;
            }
        }

        /// <summary>
        /// Populates the given permissions roles.
        /// </summary>
        /// <param name="refCollection">reference to the role or group colletion</param>
        /// <param name="roleOrGroups">An array of rows for the role or group</param>
        /// <param name="permissionType">the type of role or group (View, Add, Edit, Delete)</param>
        /// <param name="dataColumnName">Name of the column containg the data... will be different for roles and groups.</param>
        /// <remarks></remarks>
        private void setRolesOrGroups(Collection<String> refCollection, DataRow[] roleOrGroups, PermissionType permissionType, String dataColumnName)
        {
            refCollection = new Collection<String>();
            foreach (DataRow row in roleOrGroups)
            {
                if (!Convert.IsDBNull(row[m_PermissionColumn]))
                {
                    if (!Convert.IsDBNull(row[dataColumnName]))
                    {
                        if (int.Parse(row[m_PermissionColumn].ToString()) == (int)permissionType)
                        {
                            refCollection.Add(row[dataColumnName].ToString());
                        }
                    }
                }
            }
        }

        /// <summary>
        /// Sets the roles or groups.
        /// </summary>
        /// <param name="stringCollectionObject">The string collection object.</param>
        /// <param name="commaSeporatedString">The comma seporated string.</param>
        private void setRolesOrGroups(Collection<String> stringCollectionObject, string commaSeporatedString)
        {
            string[] mRoles = commaSeporatedString.Split(',');
            stringCollectionObject = new Collection<string>();
            foreach (object mRole in mRoles)
            {
                stringCollectionObject.Add(mRole.ToString());
            }

        }

        /// <summary>
        /// Populates all permissions.
        /// </summary>
        /// <param name="SecurityRows">DataRowCollection containing all derived roles for all permissions</param>
        protected void PopulatePermissions(DataRowCollection SecurityRows)
        {
            throw new NotImplementedException();
        }
    }
}

File: VB/GrowthWare.Framework/Model/Profiles/Base/MGroupRolePermissionSecurity.vb
﻿Imports GrowthWare.Framework.Model.Enumerations
Imports GrowthWare.Framework.Model.Profiles.Interfaces
Imports System.Collections.ObjectModel
Imports System.Globalization

Namespace Model.Profiles.Base
    Public MustInherit Class MGroupRolePermissionSecurity
        Inherits MProfile
        Implements IMGroupRolePermissionSecurity

        Private m_DerivedAddRoles As Collection(Of String) = New Collection(Of String)
        Private m_DerivedDeleteRoles As Collection(Of String) = New Collection(Of String)
        Private m_DerivedEditRoles As Collection(Of String) = New Collection(Of String)
        Private m_DerivedViewRoles As Collection(Of String) = New Collection(Of String)

        Private m_AssignedAddRoles As Collection(Of String) = New Collection(Of String)
        Private m_AssignedDeleteRoles As Collection(Of String) = New Collection(Of String)
        Private m_AssignedEditRoles As Collection(Of String) = New Collection(Of String)
        Private m_AssignedViewRoles As Collection(Of String) = New Collection(Of String)

        Private m_AddGroups As Collection(Of String) = New Collection(Of String)
        Private m_DeleteGroups As Collection(Of String) = New Collection(Of String)
        Private m_EditGroups As Collection(Of String) = New Collection(Of String)
        Private m_ViewGroups As Collection(Of String) = New Collection(Of String)
        Private m_PermissionColumn As String = "PERMISSIONS_SEQ_ID"
        Private m_RoleColumn As String = "Role"
        Private m_GroupColumn As String = "Group"

        ''' <summary>
        ''' Initialize orverloads and calles mybase.init to will populate the Add, Delete, Edit, and View role properties.
        ''' </summary>
        ''' <param name="detailDatarow">A data row that contains base information</param>
        ''' <param name="derivedRoles">An array of data rows that must contain two columns ("PERMISSIONS_SEQ_ID","ROLE")</param>
        ''' <param name="assignedRoles">An array of data rows that must contain two columns ("PERMISSIONS_SEQ_ID","ROLE")</param>
        ''' <param name="groups">An array of data rows that must contain two columns ("PERMISSIONS_SEQ_ID","ROLE")</param>
        ''' <remarks></remarks>
        Friend Overloads Sub Initialize(ByRef detailDataRow As DataRow, ByVal derivedRoles() As DataRow, ByVal assignedRoles() As DataRow, ByVal groups() As DataRow)
            MyBase.Initialize(detailDataRow)
            setRoleOrGroup(m_DerivedAddRoles, derivedRoles, RoleType.AddRole, m_RoleColumn)
            setRoleOrGroup(m_DerivedDeleteRoles, derivedRoles, RoleType.DeleteRole, m_RoleColumn)
            setRoleOrGroup(m_DerivedEditRoles, derivedRoles, RoleType.EditRole, m_RoleColumn)
            setRoleOrGroup(m_DerivedViewRoles, derivedRoles, RoleType.ViewRole, m_RoleColumn)
            If Not assignedRoles Is Nothing Then
                setRoleOrGroup(m_AssignedAddRoles, assignedRoles, RoleType.AddRole, m_RoleColumn)
                setRoleOrGroup(m_AssignedDeleteRoles, assignedRoles, RoleType.DeleteRole, m_RoleColumn)
                setRoleOrGroup(m_AssignedEditRoles, assignedRoles, RoleType.EditRole, m_RoleColumn)
                setRoleOrGroup(m_AssignedViewRoles, assignedRoles, RoleType.ViewRole, m_RoleColumn)
            End If
            If Not groups Is Nothing Then
                setRoleOrGroup(m_AddGroups, groups, RoleType.AddRole, m_GroupColumn)
                setRoleOrGroup(m_DeleteGroups, groups, RoleType.DeleteRole, m_GroupColumn)
                setRoleOrGroup(m_EditGroups, groups, RoleType.EditRole, m_GroupColumn)
                setRoleOrGroup(m_ViewGroups, groups, RoleType.ViewRole, m_GroupColumn)
            End If

        End Sub

        ''' <summary>
        ''' Return roles associated with the "Add" permission.
        ''' </summary>
        Public ReadOnly Property AssignedAddRoles() As Collection(Of String) Implements IMGroupRolePermissionSecurity.AssignedAddRoles
            Get
                Return m_AssignedAddRoles
            End Get
        End Property

        ''' <summary>
        ''' Return roles associated with the "Add" permission.
        ''' </summary>
        Public ReadOnly Property DerivedAddRoles() As Collection(Of String) Implements IMGroupRolePermissionSecurity.DerivedAddRoles
            Get
                Return m_DerivedAddRoles
            End Get
        End Property

        ''' <summary>
        ''' Return assigned roles associated with the "Delete" permission.
        ''' </summary>
        Public ReadOnly Property AssignedDeleteRoles() As Collection(Of String) Implements IMGroupRolePermissionSecurity.AssignedDeleteRoles
            Get
                Return m_AssignedDeleteRoles
            End Get
        End Property

        ''' <summary>
        ''' Return roles associated with the "Delete" permission.
        ''' </summary>
        Public ReadOnly Property DerivedDeleteRoles() As Collection(Of String) Implements IMGroupRolePermissionSecurity.DerivedDeleteRoles
            Get
                Return m_DerivedDeleteRoles
            End Get
        End Property

        ''' <summary>
        ''' Return assigned roles associated with the "Edit" permission.
        ''' </summary>
        Public ReadOnly Property AssignedEditRoles() As Collection(Of String) Implements IMGroupRolePermissionSecurity.AssignedEditRoles
            Get
                Return m_AssignedEditRoles
            End Get
        End Property

        ''' <summary>
        ''' Return roles associated with the "Edit" permission.
        ''' </summary>
        Public ReadOnly Property DerivedEditRoles() As Collection(Of String) Implements IMGroupRolePermissionSecurity.DerivedEditRoles
            Get
                Return m_DerivedEditRoles
            End Get
        End Property

        ''' <summary>
        ''' Return assigned roles associated with the "View" permission.
        ''' </summary>
        Public ReadOnly Property AssignedViewRoles() As Collection(Of String) Implements IMGroupRolePermissionSecurity.AssignedViewRoles
            Get
                Return m_AssignedViewRoles
            End Get
        End Property

        ''' <summary>
        ''' Return roles associated with the "View" permission.
        ''' </summary>
        Public ReadOnly Property DerivedViewRoles() As Collection(Of String) Implements IMGroupRolePermissionSecurity.DerivedViewRoles
            Get
                Return m_DerivedViewRoles
            End Get
        End Property

        Public ReadOnly Property AddGroups As Collection(Of String) Implements IMGroupRolePermissionSecurity.AddGroups
            Get
                Return m_AddGroups
            End Get
        End Property

        Public ReadOnly Property DeleteGroups As Collection(Of String) Implements IMGroupRolePermissionSecurity.DeleteGroups
            Get
                Return m_DeleteGroups
            End Get
        End Property

        Public ReadOnly Property EditGroups As Collection(Of String) Implements IMGroupRolePermissionSecurity.EditGroups
            Get
                Return m_EditGroups
            End Get
        End Property

        Public ReadOnly Property ViewGroups As Collection(Of String) Implements IMGroupRolePermissionSecurity.ViewGroups
            Get
                Return m_ViewGroups
            End Get
        End Property

        ''' <summary>
        ''' Represents the permission column name.
        ''' </summary>
        Public Property PermissionColumn() As String
            Get
                Return m_PermissionColumn
            End Get
            Set(ByVal value As String)
                m_PermissionColumn = value
                If Not String.IsNullOrEmpty(m_PermissionColumn) Then m_PermissionColumn = m_PermissionColumn.Trim()
            End Set
        End Property

        ''' <summary>
        ''' Represents the role column name.
        ''' </summary>
        Public Property RoleColumn() As String
            Get
                Return m_RoleColumn
            End Get
            Set(ByVal value As String)
                m_RoleColumn = value
                If Not String.IsNullOrEmpty(m_RoleColumn) Then m_RoleColumn = m_RoleColumn.Trim()
            End Set
        End Property

        ''' <summary>
        ''' Will set the collection of roles given a comma separated string of roles.
        ''' </summary>
        ''' <param name="CommaSeparatedRoles">String of comma separated roles</param>
        Public Sub SetAssignedRoles(ByVal commaSeparatedRoles As String, ByVal permission As PermissionType)
            Select Case permission
                Case PermissionType.Add
                    setRolesOrGroups(m_AssignedAddRoles, commaSeparatedRoles)
                Case PermissionType.Delete
                    setRolesOrGroups(m_AssignedDeleteRoles, commaSeparatedRoles)
                Case PermissionType.Edit
                    setRolesOrGroups(m_AssignedEditRoles, commaSeparatedRoles)
                Case PermissionType.View
                    setRolesOrGroups(m_AssignedViewRoles, commaSeparatedRoles)
                Case Else
            End Select
        End Sub

        ''' <summary>
        ''' Will set the collection of groups given a comma separated string of groups.
        ''' </summary>
        ''' <param name="CommaSeparatedGroups">String of comma separated groups</param>
        Public Sub SetGroups(ByVal commaSeparatedGroups As String, ByVal permission As PermissionType)
            Select Case permission
                Case PermissionType.Add
                    setRolesOrGroups(m_AddGroups, commaSeparatedGroups)
                Case PermissionType.Delete
                    setRolesOrGroups(m_DeleteGroups, commaSeparatedGroups)
                Case PermissionType.Edit
                    setRolesOrGroups(m_EditGroups, commaSeparatedGroups)
                Case PermissionType.View
                    setRolesOrGroups(m_ViewGroups, commaSeparatedGroups)
                Case Else
                    'Do nothing
            End Select
        End Sub

        ''' <summary>
        ''' Converts the collection of AssignedRoles to a comma separated string.
        ''' </summary>
        ''' <returns>String</returns>
        Public Function GetCommaSeparatedAssignedRoles(ByVal permission As PermissionType) As String
            Select Case permission
                Case PermissionType.Add
                    Return getCommaSeportatedString(m_AssignedAddRoles)
                Case PermissionType.Delete
                    Return getCommaSeportatedString(m_AssignedDeleteRoles)
                Case PermissionType.Edit
                    Return getCommaSeportatedString(m_AssignedEditRoles)
                Case PermissionType.View
                    Return getCommaSeportatedString(m_AssignedViewRoles)
                Case Else
                    Return getCommaSeportatedString(m_AssignedViewRoles)
            End Select
        End Function

        ''' <summary>
        ''' Converts the collection of AssignedGroups to a comma separated string.
        ''' </summary>
        ''' <returns>String</returns>
        Public Function GetCommaSeparatedGroups(ByVal permission As PermissionType) As String
            Select Case permission
                Case PermissionType.Add
                    Return getCommaSeportatedString(m_AddGroups)
                Case PermissionType.Delete
                    Return getCommaSeportatedString(m_DeleteGroups)
                Case PermissionType.Edit
                    Return getCommaSeportatedString(m_EditGroups)
                Case PermissionType.View
                    Return getCommaSeportatedString(m_ViewGroups)
                Case Else
                    Return getCommaSeportatedString(m_ViewGroups)
            End Select
        End Function

        ' ''' <summary>
        ' ''' Converts the collection of DerivedRoles to a comma separated string.
        ' ''' </summary>
        ' ''' <returns>String</returns>
        'Public Function GetCommaSeparatedDerivedRoles() As String
        '	Return Me.getCommaSeportatedString(m_DerivedRoles)
        'End Function


#Region "Private Methods"
        ''' <summary>
        ''' Populates the given permissions roles.
        ''' </summary>
        ''' <param name="refCollection">reference to the role or group colletion</param>
        ''' <param name="roleOrGroups">An array of rows for the role or group</param>
        ''' <param name="permissionType">the type of role or group (View, Add, Edit, Delete)</param>
        ''' <param name="dataColumnName">Name of the column containg the data... will be different for roles and groups.</param>
        ''' <remarks></remarks>
        Private Sub setRoleOrGroup(ByRef refCollection As Collection(Of String), ByVal roleOrGroups() As DataRow, ByVal permissionType As PermissionType, ByVal dataColumnName As String)
            refCollection = New Collection(Of String)
            Dim row As DataRow
            For Each row In roleOrGroups
                If Not IsDBNull(row(m_PermissionColumn)) Then
                    If (Not IsDBNull(row(dataColumnName))) Then
                        If Integer.Parse(row(m_PermissionColumn).ToString(), CultureInfo.InvariantCulture) = CType(permissionType, Integer) Then
                            refCollection.Add(row(dataColumnName).ToString())
                        End If
                    End If
                End If
            Next row
        End Sub

        ''' <summary>
        ''' Sets the assigned roles or groups.
        ''' </summary>
        ''' <param name="StringCollectionObject">The collection of roles or groups that need to be set</param>
        ''' <param name="CommaSeparatedString">A comma separated list of roles or groups 'you, me' as an example</param>
        Private Shared Sub setRolesOrGroups(ByRef stringCollectionObject As Collection(Of String), ByVal commaSeparatedString As String)
            Dim mRoles() As String = commaSeparatedString.Split(",")
            stringCollectionObject = New Collection(Of String)
            For Each mRole In mRoles
                stringCollectionObject.Add(mRole.ToString())
            Next
        End Sub

        Private Shared Function getCommaSeportatedString(ByRef collectionOfStrings As Collection(Of String)) As String
            Dim mRetValue As String = String.Empty
            If Not collectionOfStrings Is Nothing Then
                If collectionOfStrings.Count > 0 Then
                    For Each item In collectionOfStrings
                        mRetValue += item.ToString() + ","
                    Next
                End If
            End If
            If mRetValue.Length > 0 Then
                mRetValue = mRetValue.Substring(0, mRetValue.Length - 1)
            End If
            Return mRetValue
        End Function

#End Region
    End Class
End Namespace
